src/set_theory/game/nim.lean

```diff
@@ -0,0 +1,257 @@
+/-
+Copyright (c) 2020 Fox Thomson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fox Thomson
+-/
+import set_theory.game.impartial
+import set_theory.ordinal
+import data.set
+import logic.basic
+
+universes u v
+
+/-!
+# Nim and the Sprague-Grundy theorem
+
+This file contains the defition for nim for any ordinal `O`. In the game of `nim O₁` both players
+may move to `nim O₂` for any `O₂ < O₁`.
+We also define a Grundy number for an impartial game `G` and prove the Sprague-Grundy theorem, that
+`G` is equivalent to `nim (Grundy_value G)`.
+-/
+
+open pgame
+
+local infix ` ≈ ` := equiv
+
+/-- The defineition of nim, nim O is the game where each player can move to nim O' where O' < O
+    If O is finite then this game can be veiwed as a pile of stones where each player can take a
+  positie number of stones from it on their turn -/
+noncomputable def nim : ordinal → pgame
+| O₁ := ⟨ O₁.out.α, O₁.out.α,
+          λ O₂, have hwf : (ordinal.typein O₁.out.r O₂) < O₁, from begin nth_rewrite_rhs 0 ←ordinal.type_out O₁, exact ordinal.typein_lt_type _ _ end, nim (ordinal.typein O₁.out.r O₂),
+          λ O₂, have hwf : (ordinal.typein O₁.out.r O₂) < O₁, from begin nth_rewrite_rhs 0 ←ordinal.type_out O₁, exact ordinal.typein_lt_type _ _ end, nim (ordinal.typein O₁.out.r O₂)⟩
+using_well_founded {dec_tac := tactic.assumption}
+
+namespace nim
+
+lemma nim_def (O : ordinal) : nim O = pgame.mk
+          O.out.α O.out.α
+          (λ O₂, nim (ordinal.typein O.out.r O₂))
+          (λ O₂, nim (ordinal.typein O.out.r O₂)) :=
+by rw nim
+
+lemma nim_wf_lemma {O₁ : ordinal} (O₂ : O₁.out.α) : (ordinal.typein O₁.out.r O₂) < O₁ :=
+begin
+  nth_rewrite_rhs 0 ← ordinal.type_out O₁, exact ordinal.typein_lt_type _ _
+end
+
+@[instance] lemma nim_impartial : ∀ (O : ordinal), impartial (nim O)
```

Comment:
```suggestion
@[instance] lemma nim_impartial : ∀ (O : ordinal), impartial (nim O)
```
You can just write `instance nim_impartial` instead of `@[instance] lemma`.